File: backend/app/api/v1/endpoints/grades.py
from typing import List
from datetime import date

from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.orm import Session
from sqlalchemy import func

from app.api.deps import get_db, get_current_user, get_current_teacher
from app.models.user import User, UserRole
from app.models.grade import Grade
from app.models.student import StudentProfile
from app.models.teacher import TeacherProfile, Subject
from app.schemas.grade import GradeCreate, GradeRead, GradeUpdate, GradeSummary

router = APIRouter()


@router.get("", response_model=List[GradeRead])
async def get_grades(
    student_id: int = None,
    subject_id: int = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Get grades with optional filters."""
    query = db.query(Grade)

    if student_id:
        query = query.filter(Grade.student_id == student_id)
    if subject_id:
        query = query.filter(Grade.subject_id == subject_id)

    grades = query.order_by(Grade.date.desc()).all()
    return grades


@router.get("/my", response_model=List[GradeRead])
async def get_my_grades(
    subject_id: int = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Get current student's grades."""
    if current_user.role != UserRole.STUDENT:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only students can view their grades",
        )

    student_profile = (
        db.query(StudentProfile)
        .filter(StudentProfile.user_id == current_user.id)
        .first()
    )
    if not student_profile:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Student profile not found",
        )

    query = db.query(Grade).filter(Grade.student_id == student_profile.id)

    if subject_id:
        query = query.filter(Grade.subject_id == subject_id)

    grades = query.order_by(Grade.date.desc()).all()
    return grades


@router.get("/summary", response_model=List[GradeSummary])
async def get_grade_summary(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Get grade summary by subject for current student."""
    if current_user.role != UserRole.STUDENT:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only students can view grade summary",
        )

    student_profile = (
        db.query(StudentProfile)
        .filter(StudentProfile.user_id == current_user.id)
        .first()
    )
    if not student_profile:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Student profile not found",
        )

    # Get summary by subject
    summaries = (
        db.query(
            Grade.subject_id,
            Subject.name.label("subject_name"),
            func.avg(Grade.grade_value).label("average"),
            func.count(Grade.id).label("total_grades"),
            func.max(Grade.grade_value).label("highest"),
            func.min(Grade.grade_value).label("lowest"),
        )
        .join(Subject, Grade.subject_id == Subject.id)
        .filter(Grade.student_id == student_profile.id)
        .group_by(Grade.subject_id, Subject.name)
        .all()
    )

    return [
        GradeSummary(
            subject_id=s.subject_id,
            subject_name=s.subject_name,
            average=round(s.average, 2),
            total_grades=s.total_grades,
            highest=s.highest,
            lowest=s.lowest,
        )
        for s in summaries
    ]


# Bonus points thresholds: percentage -> bonus points
GRADE_BONUS_THRESHOLDS = {
    90: 10,  # 90%+ = 10 bonus points
    80: 5,   # 80-89% = 5 bonus points
    70: 2,   # 70-79% = 2 bonus points
}


def _award_bonus_points(
    db: Session, student_id: int, grade_value: float, max_value: float
) -> int:
    """Calculate and award bonus points based on grade percentage."""
    if max_value <= 0:
        return 0

    percentage = (grade_value / max_value) * 100

    for threshold, points in sorted(GRADE_BONUS_THRESHOLDS.items(), reverse=True):
        if percentage >= threshold:
            student = (
                db.query(StudentProfile)
                .filter(StudentProfile.id == student_id)
                .first()
            )
            if student:
                student.bonus_points = (student.bonus_points or 0) + points
                db.commit()
                return points
            break
    return 0


@router.post("", response_model=GradeRead, status_code=status.HTTP_201_CREATED)
async def create_grade(
    grade_in: GradeCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_teacher),
):
    """Create a new grade (Teacher only)."""
    teacher_profile = (
        db.query(TeacherProfile)
        .filter(TeacherProfile.user_id == current_user.id)
        .first()
    )
    if not teacher_profile:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Teacher profile not found",
        )

    grade = Grade(
        **grade_in.model_dump(),
        teacher_id=teacher_profile.id,
    )
    db.add(grade)
    db.commit()
    db.refresh(grade)

    # Award bonus points for good grades
    _award_bonus_points(
        db,
        student_id=grade.student_id,
        grade_value=grade.grade_value,
        max_value=grade.max_value,
    )

    return grade


@router.put("/{grade_id}", response_model=GradeRead)
async def update_grade(
    grade_id: int,
    grade_in: GradeUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_teacher),
):
    """Update a grade (Teacher only)."""
    grade = db.query(Grade).filter(Grade.id == grade_id).first()
    if not grade:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Grade not found",
        )

    update_data = grade_in.model_dump(exclude_unset=True)
    for field, value in update_data.items():
        setattr(grade, field, value)

    db.commit()
    db.refresh(grade)
    return grade


@router.delete("/{grade_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_grade(
    grade_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_teacher),
):
    """Delete a grade (Teacher only)."""
    grade = db.query(Grade).filter(Grade.id == grade_id).first()
    if not grade:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Grade not found",
        )

    db.delete(grade)
    db.commit()
    return None

File: backend/src/routes/v1/grades.ts
import type { FastifyPluginAsync } from 'fastify';

import { TEACHER_PLUS_ROLES } from '../../lib/auth.js';
import { badRequest, forbidden, notFound } from '../../lib/errors.js';
import { awardBonusPoints } from '../../lib/gradeBonus.js';
import { assertTenGrade, tenToRatio } from '../../lib/gradingScale.js';
import { prisma } from '../../lib/prisma.js';
import { serializeGrade } from '../../lib/serializers.js';
import { parseDateOnly } from '../../lib/time.js';

const TEN_SCALE_MAX_VALUE = 10;

function parseTenScaleGrade(value: unknown, fieldName = 'grade_value'): number {
  const parsed = Number(value);
  try {
    assertTenGrade(parsed);
  } catch {
    badRequest(`${fieldName} must be an integer between 2 and 10`);
  }
  return parsed;
}

function assertTenScaleMaxValue(value: unknown): void {
  if (value === undefined || value === null) {
    return;
  }

  if (Number(value) !== TEN_SCALE_MAX_VALUE) {
    badRequest('max_value must be 10');
  }
}

const gradesRoutes: FastifyPluginAsync = async (fastify) => {
  fastify.get('', { preHandler: [fastify.authenticate] }, async (request) => {
    const query = request.query as { student_id?: string | number; subject_id?: string | number };
    const currentUser = request.currentUser!;
    const where: Record<string, unknown> = {
      ...(query.subject_id ? { subject_id: Number(query.subject_id) } : {}),
    };

    if (currentUser.role === 'student') {
      const student = await prisma.studentProfile.findUnique({ where: { user_id: currentUser.id } });
      if (!student) {
        badRequest('Student profile not found');
      }
      where.student_id = student.id;
    } else if (query.student_id) {
      where.student_id = Number(query.student_id);
    }

    const grades = await prisma.grade.findMany({
      where,
      orderBy: { date: 'desc' },
    });

    return grades.map(serializeGrade);
  });

  fastify.get('/my', { preHandler: [fastify.authenticate] }, async (request) => {
    const query = request.query as { subject_id?: string | number };
    const currentUser = request.currentUser!;

    if (currentUser.role !== 'student') {
      forbidden('Only students can view their grades');
    }

    const student = await prisma.studentProfile.findUnique({ where: { user_id: currentUser.id } });
    if (!student) {
      badRequest('Student profile not found');
    }

    const grades = await prisma.grade.findMany({
      where: {
        student_id: student.id,
        ...(query.subject_id ? { subject_id: Number(query.subject_id) } : {}),
      },
      orderBy: { date: 'desc' },
    });

    return grades.map(serializeGrade);
  });

  fastify.get('/summary', { preHandler: [fastify.authenticate] }, async (request) => {
    const currentUser = request.currentUser!;

    if (currentUser.role !== 'student') {
      forbidden('Only students can view grade summary');
    }

    const student = await prisma.studentProfile.findUnique({ where: { user_id: currentUser.id } });
    if (!student) {
      badRequest('Student profile not found');
    }

    const grades = await prisma.grade.findMany({
      where: { student_id: student.id },
      include: { subject: true },
    });

    const grouped = new Map<number, { subject_name: string; values: number[] }>();
    for (const grade of grades) {
      const entry = grouped.get(grade.subject_id) ?? {
        subject_name: grade.subject.name,
        values: [],
      };
      entry.values.push(grade.grade_value);
      grouped.set(grade.subject_id, entry);
    }

    return Array.from(grouped.entries()).map(([subject_id, value]) => {
      const total_grades = value.values.length;
      const average = total_grades > 0 ? value.values.reduce((acc, item) => acc + item, 0) / total_grades : 0;
      const highest = total_grades > 0 ? Math.max(...value.values) : 0;
      const lowest = total_grades > 0 ? Math.min(...value.values) : 0;

      return {
        subject_id,
        subject_name: value.subject_name,
        average: Math.round(average * 100) / 100,
        total_grades,
        highest,
        lowest,
      };
    });
  });

  fastify.post('', { preHandler: [fastify.requireRoles(TEACHER_PLUS_ROLES, 'Teacher access required')] }, async (request, reply) => {
    const body = request.body as {
      student_id: number;
      subject_id: number;
      grade_value: number;
      max_value?: number;
      grade_type: string;
      reference_id?: number;
      date: string;
      comment?: string;
    };

    const currentUser = request.currentUser!;
    const teacher = await prisma.teacherProfile.findUnique({ where: { user_id: currentUser.id } });
    if (!teacher) {
      badRequest('Teacher profile not found');
    }
    assertTenScaleMaxValue(body.max_value);
    const gradeValue = parseTenScaleGrade(body.grade_value);

    const grade = await prisma.grade.create({
      data: {
        student_id: body.student_id,
        subject_id: body.subject_id,
        teacher_id: teacher.id,
        grade_value: gradeValue,
        max_value: TEN_SCALE_MAX_VALUE,
        grade_type: body.grade_type,
        reference_id: body.reference_id ?? null,
        date: parseDateOnly(body.date),
        comment: body.comment ?? null,
      },
    });

    await awardBonusPoints(grade.student_id, tenToRatio(grade.grade_value) * 100);
    return reply.status(201).send(serializeGrade(grade));
  });

  fastify.put(
    '/:grade_id',
    { preHandler: [fastify.requireRoles(TEACHER_PLUS_ROLES, 'Teacher access required')] },
    async (request) => {
      const params = request.params as { grade_id: string };
      const body = request.body as {
        grade_value?: number;
        max_value?: number;
        comment?: string;
      };

      const gradeId = Number(params.grade_id);
      const grade = await prisma.grade.findUnique({ where: { id: gradeId } });
      if (!grade) {
        notFound('Grade not found');
      }
      assertTenScaleMaxValue(body.max_value);
      const nextGradeValue = body.grade_value !== undefined ? parseTenScaleGrade(body.grade_value) : undefined;

      const updated = await prisma.grade.update({
        where: { id: gradeId },
        data: {
          max_value: TEN_SCALE_MAX_VALUE,
          ...(nextGradeValue !== undefined ? { grade_value: nextGradeValue } : {}),
          ...(body.comment !== undefined ? { comment: body.comment } : {}),
        },
      });

      return serializeGrade(updated);
    }
  );

  fastify.delete(
    '/:grade_id',
    { preHandler: [fastify.requireRoles(TEACHER_PLUS_ROLES, 'Teacher access required')] },
    async (request, reply) => {
      const params = request.params as { grade_id: string };
      const gradeId = Number(params.grade_id);

      const grade = await prisma.grade.findUnique({ where: { id: gradeId } });
      if (!grade) {
        notFound('Grade not found');
      }

      await prisma.grade.delete({ where: { id: gradeId } });
      return reply.status(204).send();
    }
  );
};

export default gradesRoutes;
